lib/rust/web/src/lib.rs

```diff
@@ -71,36 +70,141 @@ pub use binding::mock::*;
+    ( $($name:ident),* $(,)?) => {
+        /// WASM-target oriented traits. Extending the possibilities of wasm-bindgen structures.
+        pub mod wasm_traits {
+            $(pub use super::$name::WasmTrait as $name;)*
+            pub use super::binding::wasm::JsCast;
+        }
+
+        /// Mock traits. Counterpart of [`wasm_traits`].
+        pub mod mock_traits {
+            $(pub use super::$name::MockTrait as $name;)*
+            pub use super::binding::mock::JsCast;
+        }
+
+        /// Both wasm and mock traits, unnamed.
+        pub mod anonymous_traits {
+            $(pub use super::$name::WasmTrait as _;)*
+            $(pub use super::$name::MockTrait as _;)*
+        }
+    };
+}
+
+gen_traits! {
+    ClosureOps,
+    DocumentOps,
+    ElementOps,
+    FunctionOps,
+    HtmlCanvasElementOps,
+    HtmlElementOps,
+    JsValueOps,
+    NodeOps,
+    ObjectOps,
+    ReflectOps,
+    WindowOps,
+}
+
 /// All traits defined in this module.
 pub mod traits {
-    pub use super::ClosureOps;
-    pub use super::DocumentOps;
-    pub use super::ElementOps;
-    pub use super::FunctionOps;
-    pub use super::HtmlCanvasElementOps;
-    pub use super::HtmlElementOps;
-    pub use super::JsCast;
-    pub use super::NodeOps;
-    pub use super::ObjectOps;
-    pub use super::ReflectOps;
-    pub use super::WindowOps;
+    pub use super::anonymous_traits::*;
+    #[cfg(not(target_arch = "wasm32"))]
+    pub use super::mock_traits::*;
+    #[cfg(target_arch = "wasm32")]
+    pub use super::wasm_traits::*;
+}
+
+
+macro_rules! ops {
+    ($(<$($arg:ident : ($($arg_tp:tt)*)),*>)? $trait:ident for $target:ident
+    trait $defs:tt
+    $(impl {$($body:tt)*})?
+    $(wasm_impl {$($wasm_body:tt)*})?
+    $(mock_impl {$($mock_body:tt)*})?
+    ) => {
+        /// [`$trait`] extensions.
+        #[allow(non_snake_case)]
+        #[allow(missing_docs)]
+        #[allow(unused_imports)]
+        pub mod $trait {
+            use super::*;
+
+            /// WASM bindings.
+            pub mod wasm {
+                use enso_prelude::*;
+                use super::binding::wasm::*;
+                use super::wasm_traits::*;
+                /// Extensions to the [`$target`] type.
+                pub trait Trait $defs
+                impl $(<$($arg: $($arg_tp)*),*>)? Trait for $target $(<$($arg),*>)?
+                    {$($($body)*)? $($($wasm_body)*)?}
+            }
+
+            /// Mock bindings.
+            pub mod mock {
+                use enso_prelude::*;
+                use super::binding::mock::*;
+                use super::mock_traits::*;
+                /// Extensions to the [`$target`] type.
+                pub trait Trait $defs
+                impl $(<$($arg: $($arg_tp)*),*>)? Trait for $target $(<$($arg),*>)?
+                    {$($($body)*)? $($($mock_body)*)?}
+            }
+            pub use self::wasm::Trait as WasmTrait;
+            pub use self::mock::Trait as MockTrait;
+        }
+    };
```

Comment:
Could you add a short description to this macro? Explaining also why we define both traits without any guards here.

An aside note, I know one of the reasons for not having "target" guards is to make sure that both codes compiles. However, I think some build script (perhaps `./run check?`) could run `cargo check` for both native platform and WASM. What do you think about it?

I did a quick research: `cargo check -p enso-gui --all-targets --target wasm32-unknonwn-unknown` works. We can add other crates with `-p` which are not a dependency of enso-gui, and we want to check for wasm32 build (parser maybe?).